### ISSUE TITLE
Retrieve preprocessing progress  during the Ludwig data preprocessing

### ISSUE THREAD

**Issue by SumanthDatta-Kony** [ludwig-ai/ludwig#4195]
Currently during preprocessing , how do we get the percentage of data or the amount of data already preprocessed. 
We have callbacks implemented in our project. Is there a way to get the details of the progress through callbacks.

If there is no way to retrieve this progress value. Can it be taken as a feature request and implemented in the future.

**Comment by w4nderlust** [comment]
Good feature request! There is currently no way to get preprocessing progress through callbacks, but we just implemented it in https://github.com/ludwig-ai/ludwig/commit/7d5754a7f.

A new `on_preprocess_progress(progress: float)` hook is now available on the `Callback` base class. It fires periodically during `build_data` with a value in `[0.0, 1.0]`. Usage:

```python
from ludwig.callbacks import Callback

class MyCallback(Callback):
    def on_preprocess_progress(self, progress, **kwargs):
        print(f"Preprocessing: {progress:.0%}")

model = LudwigModel(config, callbacks=[MyCallback()])
```

Progress is tracked at the partition level, so the granularity depends on the backend:
- **pandas/modin**: one update per feature (the whole column is one partition)
- **Ray**: one update per Dask partition per feature, fired from inside the remote worker via a Ray actor, so it works correctly in distributed mode too

It will be included in the next release. Let us know if you have any feedback!

**Comment by SumanthDatta-Kony** [comment]
Thanks for the update , will test this in the new ludwig version.

**Comment by w4nderlust** [comment]
Great news: this is now in Ludwig 0.17.6! Just implement `on_preprocess_progress` in your `Callback` subclass:

```python
from ludwig.callbacks import Callback
from ludwig.api import LudwigModel

class ProgressCallback(Callback):
    def on_preprocess_progress(self, progress: float, **kwargs):
        print(f"Preprocessing: {progress:.0%}")

model = LudwigModel(config, callbacks=[ProgressCallback()])
model.train(dataset=df)
```

`progress` is a float in `[0.0, 1.0]` that increments once per completed partition (pandas column, Dask partition, or Ray worker task). The final call is always `1.0`. Works with all backends, no extra config needed.

Full docs: https://ludwig.ai/latest/developer_guide/add_an_integration/

**Comment by SumanthDatta-Kony** [comment]
Ludwig 0.17.6 will be available for pip downloads today? Docker images are available though.

It's not yet there on pypi.org

<img width="1116" height="432" alt="Image" src="https://github.com/user-attachments/assets/d94268b7-8d61-49ce-b8b2-775d65666a39" />

**Comment by w4nderlust** [comment]
Just published on PyPI: https://pypi.org/project/ludwig/0.17.6/

`pip install ludwig==0.17.6`

**Comment by SumanthDatta-Kony** [comment]
@w4nderlust  , with this implementation i only see 0% or 100 % , no other value in between it displays. my implementation is with ray as backend. 

After it reaches 100% , it took around 5 mins to display preprocessing completed in the logs. This dataset was  a 2 million one. 


<img width="1767" height="482" alt="Image" src="https://github.com/user-attachments/assets/2237e084-cc7e-458d-9b2e-c33d12b53a82" />

**Comment by SumanthDatta-Kony** [comment]
Here is a sample program with the 2 million dataset.  With  backend type local,  it  gives only 3 values 0 % , 19% and 100%. 

[read_csv_training.py](https://github.com/user-attachments/files/29526164/read_csv_training.py)
[Callback_class.py](https://github.com/user-attachments/files/29526163/Callback_class.py)

dataset is in location 

https://drive.google.com/file/d/18LtD4RylXHpAlSdly9TaINqKJAJBuKwi/view?usp=drivesdk

<img width="422" height="686" alt="Image" src="https://github.com/user-attachments/assets/68b8fd04-5986-41ff-bf53-7c5c1d60b25a" />

**Comment by w4nderlust** [comment]
Thanks for sharing the sample, looking into it!

**Comment by w4nderlust** [comment]
Found the root cause: most feature types (binary, category, number without normalization) do their work without calling `map_partitions` internally, so the counter never incremented for them. Only features that went through `map_partitions` (a subset) produced intermediate callbacks, which is why you saw so few steps.

Fixed in **0.17.7** (just released): the increment now fires once per feature in the `build_data` loop, regardless of which engine operation the feature uses internally. You should see one step per input/output feature now.

```
pip install ludwig==0.17.7
```

Would love to hear if the granularity looks right with your 2M dataset after upgrading!

**Comment by SumanthDatta-Kony** [comment]
@w4nderlust , thanks for the implementation , it is working as expected .

<img width="635" height="822" alt="Image" src="https://github.com/user-attachments/assets/5be4c41d-ff96-4f09-a17f-f524d93d4750" />

**Comment by w4nderlust** [comment]
You are welcome!
